src/sorted_iterator.rs

```diff
@@ -54,6 +55,94 @@ impl<K: Ord, I: Iterator<Item = K>, J: Iterator<Item = K>> Iterator for Union<I,
     }
 }
 
+// An iterator with the first item pulled out.
+pub(crate) struct Peeked<I: Iterator>{
```

Comment:
Can you use the normal https://doc.rust-lang.org/std/iter/struct.Peekable.html for this? The only difference seems to be the Reverse. So having just a newtype wrapper for a Peekable might be slightly less boilerplate. Not the end of the world though.

Reply:
The problem is that in order to merge the iterators together with good complexity, I use `binary_heap` which requires the elements to implement `Ord`. Therefore, I think a wrapper of one kind or another is still necessary.

There is another subtle difference between the standard `Peekable` and this. The `Peeked` struct defined here always pre-fetches one element out of the wrapped iterator, whereas `Peekable` gets the element on demand. That is again useful to implement `Ord` that compares the iterators by the first element without hacks.

Reply:
> > Hi. Sorry about not following up earlier. It has been a bit crazy because of having my entire family at home due to corona...
> 
> Hi, hope you and your family are fine.
> 
We are doing well given the circumstances. It is just hard to get some free time for open source work.

> > So, `SortedIterator` is supposed to produce a strictly increasing sequence, not just non-decreasing. I will update the docs accordingly.
> > Your code handles that properly, so I guess it is OK.
> 
> In that case, I will be able to simplify the implementation since handling duplicate elements is not required. Will have a look into it when I have some time over the next couple of days.

Cool. Looking forward to it.

Reply:
> The problem is that in order to merge the iterators together with good complexity, I use `binary_heap` which requires the elements to implement `Ord`. Therefore, I think a wrapper of one kind or another is still necessary.
> 
Yes, some kind of wrapper is definitely needed.

> There is another subtle difference between the standard `Peekable` and this. The `Peeked` struct defined here always pre-fetches one element out of the wrapped iterator, whereas `Peekable` gets the element on demand. That is again useful to implement `Ord` that compares the iterators by the first element without hacks.

Well, the Peekable is not exactly super complex, so if you think it is the best way I am OK with that.